### PR TITLE
update to latest proto

### LINKED ISSUE #32: Combine Query and Execute to Run
The current proto buffer definition exposes two methods: `Exec` and `Query`.

```proto
service SQLFlow {
  rpc Query (Request) returns (stream RowSet);
  rpc Execute (Request) returns (stream Messages);
}
```

The client needs to decide which method to call. This would require SQL parsing at the client side, while the server already has a parser.

As a side note, we shouldn't let the user to choose which method to call because a typical SQL console doesn't distinguish `Query` and `Execute`.

### PR DESCRIPTION
**Pull request by tonyyang-svail**
fix #32

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -21,8 +21,7 @@ protoc: ## Generate python client from proto file
 	source build/grpc/bin/activate \
 	&& pip install grpcio-tools \
 	&& mkdir -p build/grpc/sqlflow/proto \
-	&& curl -o build/grpc/sqlflow/proto/sqlflow.proto \
-		https://raw.githubusercontent.com/wangkuiyi/sqlflowserver/$(SQLFLOW_VERSION)/sqlflow.proto \
+	&& cp sqlflow/proto/sqlflow.proto build/grpc/sqlflow/proto/sqlflow.proto \
```
Comment (ktong) [inline Makefile]: Does this work on CI?

File: setup.py [hunk 2]
```diff
@@ -20,12 +20,14 @@
...
     'ipython>=1.0',
+    'prettytable',
 ]
 SETUP_REQUIRED = [
     'pytest-runner'
 ]
 TEST_REQUIRED = [
-    'pytest'
+    'pytest',
+    'mock',
```
Comment (ktong) [inline setup.py]: Is this built-in mock? python has built-in mock under unittest

File: sqlflow/client.py [hunk 3]
```diff
@@ -7,67 +7,73 @@
 import sqlflow.proto.sqlflow_pb2 as pb
 import sqlflow.proto.sqlflow_pb2_grpc as pb_grpc
 
-_LOGGER = logging.getLogger(__name__)
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
```
Comment (ktong) [inline sqlflow/client.py]: It does not suggest to config this in library code

File: sqlflow/client.py [hunk 4]
```diff
@@ -7,67 +7,73 @@
...
+_LOGGER = logging.getLogger()
+
+
+class Rows:
+    def __init__(self, rows_gen):
+        rows = [r for r in rows_gen()]
```
Comment (ktong) [inline sqlflow/client.py]: It should not do heavy lift in __init__ method. Suggest move it to column_names and rows

File: sqlflow/client.py [hunk 5]
```diff
@@ -7,67 +7,73 @@
...
         Args:
-            server_url(str):    sqlflowserver url. If None, read value from environment
-                                variable SQLFLOW_SERVER
+            server_url(str):    sqlflowserver url.
 
         Raises:
             KeyError:
                 Raised if SQLFLOW_SERVER is not specified as environment variable
         """
-        if server_url is None:
```
Comment (ktong) [inline sqlflow/client.py]: Why remove reading server_url from environment variable?

**Comment by weiguoz** [comment]
LGTM

File: sqlflow/client.py (hunk 5 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Moved back.

**Comment by tonyyang-svail** [comment]
Also need to update `__main__.py` and remove `dbapi`

File: sqlflow/client.py [hunk 6]
```diff
@@ -33,41 +64,22 @@ def __init__(self, server_url=None):
...
-        Args:
-            res(sqlflow.proto.sqlflow_pb2.RunResponse.Columns)
-
-        Returns:
-            Dictionary from str to list
-        """
-        table = {"column_names": [name for name in res.table.column_names],
-                 "rows": [[cls._decode_any(a) for a in row.data] for row in res.table.rows]}
-        return table
+                assert res.WhichOneof('response') == 'head'
```
Comment (tonyyang-svail) [inline sqlflow/client.py]: I am not sure if we need this `assert`

File: sqlflow/client.py (hunk 3 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Changed to only set `_LOGGER`'s level and handler, instead of global config.

File: Makefile [hunk 7]
```diff
@@ -21,8 +21,7 @@ protoc: ## Generate python client from proto file
 	source build/grpc/bin/activate \
 	&& pip install grpcio-tools \
 	&& mkdir -p build/grpc/sqlflow/proto \
-	&& curl -o build/grpc/sqlflow/proto/sqlflow.proto \
-		https://raw.githubusercontent.com/wangkuiyi/sqlflowserver/$(SQLFLOW_VERSION)/sqlflow.proto \
+	&& cp proto/sqlflow.proto build/grpc/sqlflow/proto/sqlflow.proto \
```
Comment (tonyyang-svail) [inline Makefile]: @ktong 我最后还是用了copy. 我没有试出来直接 `python -m grpc_tools.protoc ...`

Comment (ktong) [reply Makefile]: push file in proto/sqlflow/proto/sqlflow.proto, and change -Iproto, and also remove **&& rm -rf build/grpc** with cp

File: sqlflow/client.py [hunk 8]
```diff
@@ -1,13 +1,50 @@
...
+
 import google.protobuf.wrappers_pb2 as wrapper
+from google.protobuf.timestamp_pb2 import Timestamp
 
 import sqlflow.proto.sqlflow_pb2 as pb
 import sqlflow.proto.sqlflow_pb2_grpc as pb_grpc
 
+
 _LOGGER = logging.getLogger(__name__)
+_OUT_HANDLER = logging.StreamHandler(sys.stdout)
```
Comment (ktong) [inline sqlflow/client.py]: Still have concern to set log handler here. may we move it into magic command?

File: sqlflow/client.py [hunk 9]
```diff
@@ -1,13 +1,50 @@
...
+        self._rows_gen = rows_gen
+        self._rows = None
+
+    def column_names(self):
+        return self._column_names
+
+    def rows(self):
+        if self._rows is None:
+            self._rows = [r for r in self._rows_gen()]
+        return self._rows
```
Comment (ktong) [inline sqlflow/client.py]: Try this:
for r in self._rows_gen():
  yield from r

File: sqlflow/client.py [hunk 10]
```diff
@@ -1,13 +1,50 @@
...
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        self.rows()
```
Comment (ktong) [inline sqlflow/client.py]: remove this?

File: sqlflow/client.py [hunk 11]
```diff
@@ -1,13 +1,50 @@
...
+
+        from prettytable import PrettyTable
+        table = PrettyTable(self._column_names)
+        for row in self._rows:
```
Comment (ktong) [inline sqlflow/client.py]: for row in self.rows()?

File: sqlflow/client.py [hunk 12]
```diff
@@ -33,41 +70,24 @@ def __init__(self, server_url=None):
...
-        Args:
-            res(sqlflow.proto.sqlflow_pb2.RunResponse.Columns)
-
-        Returns:
-            Dictionary from str to list
         """
-        table = {"column_names": [name for name in res.table.column_names],
-                 "rows": [[cls._decode_any(a) for a in row.data] for row in res.table.rows]}
-        return table
+        stream_response = self._stub.Run(pb.Request(sql=operation))
```
Comment (ktong) [inline sqlflow/client.py]: I like this version of execute

File: sqlflow/client.py (hunk 9 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Do we need to support `rows()` to be called twice? how about

```python
def rows(self):
    if self._rows is None:
        self._rows = []
        for r in self._rows_gen():
            self._rows.append(r)
            yield r
    else:
        for r in self._rows:
            yield r
```

Comment (ktong) [reply sqlflow/client.py]: cache it into _rows:)

File: sqlflow/client.py (hunk 10 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Done.

File: sqlflow/client.py (hunk 11 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Done.

File: sqlflow/client.py (hunk 9 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Done.

File: sqlflow/client.py (hunk 8 above)
Comment (tonyyang-svail) [reply sqlflow/client.py]: Done.

File: Makefile (hunk 7 above)
Comment (tonyyang-svail) [reply Makefile]: Done

**Pull request merged**